Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs

```diff
@@ -0,0 +1,344 @@
+            {
+                return;
+            }
+
+            _evaporationToken = new CancellationTokenSource();
+
+            // KYS to evaporate
+            Timer.Spawn(TimeSpan.FromSeconds(_evaporateTime), CheckEvaporate, _evaporationToken.Token);
+        }
+
+        private void UpdateAppearance()
+        {
+            if (Owner.Deleted)
+            {
+                return;
+            }
+            // Opacity based on level of fullness to overflow
+            // Hard-cap lower bound for visibility reasons
+            var volumeScale = ((float)CurrentVolume / OverflowVolume) * 0.75f + 0.25f;
+            var cappedScale = Math.Min(1.0f, volumeScale);
+            // Color based on the underlying solutioncomponent
+            Color newColor;
+            if (_recolor)
+            {
+                newColor = new Color(
+                    _contents.SubstanceColor.R,
+                    _contents.SubstanceColor.G,
+                    _contents.SubstanceColor.B,
+                    cappedScale);
+            }
+            else
+            {
+                newColor = new Color(
+                    _spriteComponent.Color.R,
+                    _spriteComponent.Color.G,
+                    _spriteComponent.Color.B,
+                    cappedScale);
+            }
+
+            _spriteComponent.Color = newColor;
+
+            _spriteComponent.Dirty();
+
+        }
+
+        /// <summary>
+        /// Will overflow this entity to neighboring entities if required
+        /// </summary>
+        private void CheckOverflow()
+        {
+            if (CurrentVolume <= _overflowVolume)
+            {
+                return;
+            }
+
+            // Essentially:
+            // Spill at least 1 solution to each neighbor (so most of the time each puddle is getting 1 max)
+            // If there's no puddle at the neighbor then add one.
+
+            // Setup
+            // If there's more neighbors to spill to then there are reagents to go around (coz integers)
+            var overflowAmount = CurrentVolume - OverflowVolume;
+
+            var neighborPuddles = new List<IEntity>(8);
+
+            // Will overflow to each neighbor; if it already has a puddle entity then add to that
+
+            foreach (var direction in RandomDirections())
+            {
+                // Can't spill < 1 reagent so stop overflowing
+                if (neighborPuddles.Count == overflowAmount)
+                {
+                    break;
+                }
+
+                // If we found an existing puddle on that tile then we don't need to spawn a new one
+                var noSpawn = false;
+
+                foreach (var entity in _snapGrid.GetInDir(direction))
+                {
+                    // Don't overflow to walls
+                    if (entity.TryGetComponent(out CollidableComponent collidableComponent) &&
+                        collidableComponent.CollisionLayer == 1)
```

Comment:
Please use collision groups defined in `Content.Shared.Physics.CollisionGroup` instead.